numpy/ma/core.py

```diff
@@ -3344,17 +3346,35 @@ def __setitem__(self, indx, value):
             _mask[indx] = mindx
         return
 
-    def __setattr__(self, attr, value):
-        super(MaskedArray, self).__setattr__(attr, value)
-        if attr == 'dtype' and self._mask is not nomask:
-            self._mask = self._mask.view(make_mask_descr(value), ndarray)
-            # Try to reset the shape of the mask (if we don't have a void)
-            # This raises a ValueError if the dtype change won't work
+    # Define so that we can overwrite the setter.
+    @property
+    def dtype(self):
+        return super(MaskedArray, self).dtype
+
+    @dtype.setter
+    def dtype(self, dtype):
+        super(MaskedArray, type(self)).dtype.__set__(self, dtype)
+        if self._mask is not nomask:
+            self._mask = self._mask.view(make_mask_descr(dtype), ndarray)
+            # Try to reset the shape of the mask (if we don't have a void).
+            # This raises a ValueError if the dtype change won't work.
             try:
                 self._mask.shape = self.shape
             except (AttributeError, TypeError):
                 pass
 
+    @property
+    def shape(self):
+        return super(MaskedArray, self).shape
```

Comment:
Pretty sure this needs to be `super(MaskedArray, type(self)).__get__(self, type(self))`, unless there's something weird about base classes in C

Reply:
Ok, seems there's something weird about base classes. If this were shadowing another `property` (rather than a `getset_descriptor`) , you'd need what I have above.

Reply:
I don't think that's true - I;m fairly sure one can shadow a property getter as I have it above, just not the setter.